### ISSUE TITLE
ZODB.POSException.ReadConflictError: '\x0e\x15\x8a\t\x83\xaf\x1e\xb5'

### ISSUE THREAD

**Issue by tisto** [plone/plone.restapi#470]
Upgrading the plone.restapi buildout from Plone 5.1b3 to 5.1b4 leads to a ZODB.POSException when running the tests:

```
$ bin/test
Running plone.restapi.testing.PloneRestApiATLayer:Functional tests:
  Set up plone.testing.zca.LayerCleanup in 0.000 seconds.
  Set up plone.testing.z2.Startup in 0.422 seconds.
  Set up plone.app.testing.layers.PloneFixture in 20.745 seconds.
  Set up plone.restapi.testing.PloneRestApiATLayer in 2.362 seconds.
  Set up plone.testing.z2.ZServer in 0.502 seconds.
  Set up plone.restapi.testing.PloneRestApiATLayer:Functional in 0.000 seconds.
  Running:

  Ran 8 tests with 0 failures and 0 errors in 6.787 seconds.
Running plone.restapi.testing.PloneRestApiATLayer:Integration tests:
  Tear down plone.restapi.testing.PloneRestApiATLayer:Functional in 0.000 seconds.
  Tear down plone.testing.z2.ZServer in 5.176 seconds.
  Set up plone.restapi.testing.PloneRestApiATLayer:Integration in 0.000 seconds.
  Running:

  Ran 88 tests with 0 failures and 0 errors in 30.821 seconds.
Running plone.restapi.testing.PloneRestApiDXLayer:Functional tests:
  Tear down plone.restapi.testing.PloneRestApiATLayer:Integration in 0.000 seconds.
  Tear down plone.restapi.testing.PloneRestApiATLayer in 0.018 seconds.
  Set up plone.app.event.testing.PAEventLayer in 2.349 seconds.
  Set up plone.app.contenttypes.testing.PloneAppContenttypes in 2.918 seconds.
  Set up plone.restapi.testing.PloneRestApiDXLayer in 4.478 seconds.
  Set up plone.testing.z2.ZServer in 0.507 seconds.
  Set up plone.restapi.testing.PloneRestApiDXLayer:Functional in 0.000 seconds.
  Running:
    76/316 (24.1%)Traceback (most recent call last):
  File "bin/test", line 292, in <module>
    '--test-path', '/Users/timo/workspace/plone/plone.restapi/src',
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/__init__.py", line 35, in run
    failed = run_internal(defaults, args, script_parts=script_parts)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/__init__.py", line 48, in run_internal
    runner.run()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/runner.py", line 138, in run
    self.run_tests()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/runner.py", line 219, in run_tests
    setup_layers, self.failures, self.errors)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/runner.py", line 376, in run_layer
    return run_tests(options, tests, layer_name, failures, errors)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/runner.py", line 308, in run_tests
    test(result)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 393, in __call__
    return self.run(*args, **kwds)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 370, in run
    result.stopTest(self)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/runner.py", line 736, in stopTest
    self.testTearDown()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.testing-3.9.7-py2.7.egg/zope/testing/testrunner/runner.py", line 688, in testTearDown
    layer.testTearDown()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.app.testing-5.0.7-py2.7.egg/plone/app/testing/layers.py", line 278, in testTearDown
    self.tearDownEnvironment(self['portal'])
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.app.testing-5.0.7-py2.7.egg/plone/app/testing/layers.py", line 316, in tearDownEnvironment
    if cache and getattr(cache, '_cacheId', None):
  File "/Users/timo/workspace/plone/plone.restapi/eggs/ZODB-5.2.4-py2.7.egg/ZODB/Connection.py", line 796, in setstate
    p, serial = self._storage.load(oid)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/ZODB-5.2.4-py2.7.egg/ZODB/mvccadapter.py", line 146, in load
    raise POSException.ReadConflictError(repr(oid))
ZODB.POSException.ReadConflictError: '\x0e\x15\x8a\t\x83\xaf\x1e\xb5'
```

**Comment by tisto** [comment]
Travis currently fails to install ZEO 5.1.0: 

https://travis-ci.org/plone/plone.restapi/jobs/323139336

We have to fix this issue first I guess...

**Comment by tisto** [comment]
Travis issue has been fixed. Problem is reproducible on Travis: 

https://travis-ci.org/plone/plone.restapi/jobs/323149842

**Comment by tisto** [comment]
I removed test_documentation.py and run bin/test. Then tests pass but I still see an error trace:

```
$ bin/test &
[1] 43499
Timos-MBP:plone.restapi timo$ Removing stale bytecode file /Users/timo/workspace/plone/plone.restapi/src/plone/restapi/tests/test_documentation.pyc
Running plone.restapi.testing.PloneRestApiATLayer:Functional tests:
  Set up plone.testing.zca.LayerCleanup in 0.000 seconds.
  Set up plone.testing.z2.Startup in 0.398 seconds.
  Set up plone.app.testing.layers.PloneFixture in 17.271 seconds.
  Set up plone.restapi.testing.PloneRestApiATLayer in 2.726 seconds.
  Set up plone.testing.z2.ZServer in 0.504 seconds.
  Set up plone.restapi.testing.PloneRestApiATLayer:Functional in 0.000 seconds.
  Running:

  Ran 8 tests with 0 failures and 0 errors in 3.914 seconds.
Running plone.restapi.testing.PloneRestApiATLayer:Integration tests:
  Tear down plone.restapi.testing.PloneRestApiATLayer:Functional in 0.000 seconds.
  Tear down plone.testing.z2.ZServer in 5.504 seconds.
  Set up plone.restapi.testing.PloneRestApiATLayer:Integration in 0.000 seconds.
  Running:

  Ran 88 tests with 0 failures and 0 errors in 15.127 seconds.
Running plone.restapi.testing.PloneRestApiDXLayer:Functional tests:
  Tear down plone.restapi.testing.PloneRestApiATLayer:Integration in 0.000 seconds.
  Tear down plone.restapi.testing.PloneRestApiATLayer in 0.006 seconds.
  Set up plone.app.event.testing.PAEventLayer in 1.110 seconds.
  Set up plone.app.contenttypes.testing.PloneAppContenttypes in 1.828 seconds.
  Set up plone.restapi.testing.PloneRestApiDXLayer in 1.902 seconds.
  Set up plone.testing.z2.ZServer in 0.502 seconds.
  Set up plone.restapi.testing.PloneRestApiDXLayer:Functional in 0.000 seconds.
  Running:
    168/246 (68.3%)  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 774, in __bootstrap
    self.__bootstrap_inner()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.testing-4.3.0-py2.7.egg/plone/testing/z2.py", line 1053, in runner
    asyncore.poll(self.timeout, socket_map)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asyncore.py", line 156, in poll
    read(obj)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asyncore.py", line 83, in read
    obj.handle_read_event()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asyncore.py", line 449, in handle_read_event
    self.handle_read()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asynchat.py", line 165, in handle_read
    self.found_terminator()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/medusa/http_server.py", line 531, in found_terminator
    h.handle_request (r)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/HTTPServer.py", line 162, in handle_request
    self.continue_request(sin,request)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/HTTPServer.py", line 276, in continue_request
    request.channel.work()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/HTTPServer.py", line 388, in work
    handle(module_name, request, response)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/PubCore/__init__.py", line 24, in handle
    return apply(_handle, args, kw)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.testing-4.3.0-py2.7.egg/plone/testing/z2.py", line 947, in threadlessPubCoreHandler
    publish_module(name, request=request, response=response)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZPublisher/Publish.py", line 455, in publish_module
    environ, debug, request, response)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZPublisher/Publish.py", line 249, in publish_module_standard
    response = publish(request, module_name, after_list, debug=debug)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZPublisher/Publish.py", line 143, in publish
    notify(PubBeforeCommit(request))
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.event-3.5.2-py2.7.egg/zope/event/__init__.py", line 31, in notify
    subscriber(event)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.component-4.3.0-py2.7.egg/zope/component/event.py", line 27, in dispatch
    component_subscribers(event, None)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.component-4.3.0-py2.7.egg/zope/component/_api.py", line 139, in subscribers
    return sitemanager.subscribers(objects, interface)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.interface-4.1.0-py2.7-macosx-10.12-x86_64.egg/zope/interface/registry.py", line 328, in subscribers
    return self.adapters.subscribers(objects, provided)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.interface-4.1.0-py2.7-macosx-10.12-x86_64.egg/zope/interface/adapter.py", line 601, in subscribers
    subscription(*objects)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.transformchain-1.2.1-py2.7.egg/plone/transformchain/zpublisher.py", line 85, in applyTransformOnSuccess
    transformed = applyTransform(event.request)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.transformchain-1.2.1-py2.7.egg/plone/transformchain/zpublisher.py", line 74, in applyTransform
    transformed = transformer(request, result, encoding)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.transformchain-1.2.1-py2.7.egg/plone/transformchain/transformer.py", line 49, in __call__
    newResult = handler.transformIterable(result, encoding)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.protect-3.0.24-py2.7.egg/plone/protect/auto.py", line 182, in transformIterable
    if not self.check():
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.protect-3.0.24-py2.7.egg/plone/protect/auto.py", line 207, in check
    return self._check()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.protect-3.0.24-py2.7.egg/plone/protect/auto.py", line 279, in _check
    traceback.print_stack(),
    169/246 (68.7%)  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 774, in __bootstrap
    self.__bootstrap_inner()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.testing-4.3.0-py2.7.egg/plone/testing/z2.py", line 1053, in runner
    asyncore.poll(self.timeout, socket_map)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asyncore.py", line 156, in poll
    read(obj)
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asyncore.py", line 83, in read
    obj.handle_read_event()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asyncore.py", line 449, in handle_read_event
    self.handle_read()
  File "/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7/lib/python2.7/asynchat.py", line 165, in handle_read
    self.found_terminator()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/medusa/http_server.py", line 531, in found_terminator
    h.handle_request (r)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/HTTPServer.py", line 162, in handle_request
    self.continue_request(sin,request)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/HTTPServer.py", line 276, in continue_request
    request.channel.work()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/HTTPServer.py", line 388, in work
    handle(module_name, request, response)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZServer/PubCore/__init__.py", line 24, in handle
    return apply(_handle, args, kw)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.testing-4.3.0-py2.7.egg/plone/testing/z2.py", line 947, in threadlessPubCoreHandler
    publish_module(name, request=request, response=response)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZPublisher/Publish.py", line 455, in publish_module
    environ, debug, request, response)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZPublisher/Publish.py", line 249, in publish_module_standard
    response = publish(request, module_name, after_list, debug=debug)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/Zope2-2.13.26-py2.7.egg/ZPublisher/Publish.py", line 143, in publish
    notify(PubBeforeCommit(request))
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.event-3.5.2-py2.7.egg/zope/event/__init__.py", line 31, in notify
    subscriber(event)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.component-4.3.0-py2.7.egg/zope/component/event.py", line 27, in dispatch
    component_subscribers(event, None)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.component-4.3.0-py2.7.egg/zope/component/_api.py", line 139, in subscribers
    return sitemanager.subscribers(objects, interface)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.interface-4.1.0-py2.7-macosx-10.12-x86_64.egg/zope/interface/registry.py", line 328, in subscribers
    return self.adapters.subscribers(objects, provided)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/zope.interface-4.1.0-py2.7-macosx-10.12-x86_64.egg/zope/interface/adapter.py", line 601, in subscribers
    subscription(*objects)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.transformchain-1.2.1-py2.7.egg/plone/transformchain/zpublisher.py", line 85, in applyTransformOnSuccess
    transformed = applyTransform(event.request)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.transformchain-1.2.1-py2.7.egg/plone/transformchain/zpublisher.py", line 74, in applyTransform
    transformed = transformer(request, result, encoding)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.transformchain-1.2.1-py2.7.egg/plone/transformchain/transformer.py", line 49, in __call__
    newResult = handler.transformIterable(result, encoding)
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.protect-3.0.24-py2.7.egg/plone/protect/auto.py", line 182, in transformIterable
    if not self.check():
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.protect-3.0.24-py2.7.egg/plone/protect/auto.py", line 207, in check
    return self._check()
  File "/Users/timo/workspace/plone/plone.restapi/eggs/plone.protect-3.0.24-py2.7.egg/plone/protect/auto.py", line 279, in _check
    traceback.print_stack(),

  Ran 246 tests with 0 failures and 0 errors in 1 minutes 24.056 seconds.
Running plone.restapi.testing.PloneRestApiDXLayer:Integration tests:
  Tear down plone.restapi.testing.PloneRestApiDXLayer:Functional in 0.000 seconds.
  Tear down plone.testing.z2.ZServer in 5.502 seconds.
  Set up plone.restapi.testing.PloneRestApiDXLayer:Integration in 0.000 seconds.
  Running:

  Ran 285 tests with 0 failures and 0 errors in 56.902 seconds.
Running plone.restapi.testing.PloneRestApiDXPAMLayer:Functional tests:
  Tear down plone.restapi.testing.PloneRestApiDXLayer:Integration in 0.000 seconds.
  Tear down plone.restapi.testing.PloneRestApiDXLayer in 0.007 seconds.
  Set up plone.restapi.testing.PloneRestApiDXPAMLayer in 2.887 seconds.
  Set up plone.testing.z2.ZServer in 0.505 seconds.
  Set up plone.restapi.testing.PloneRestApiDXPAMLayer:Functional in 0.000 seconds.
  Running:

  Ran 7 tests with 0 failures and 0 errors in 5.187 seconds.
Running plone.restapi.testing.PloneRestApiDXPAMLayer:Integration tests:
  Tear down plone.restapi.testing.PloneRestApiDXPAMLayer:Functional in 0.000 seconds.
  Tear down plone.testing.z2.ZServer in 5.352 seconds.
  Set up plone.restapi.testing.PloneRestApiDXPAMLayer:Integration in 0.000 seconds.
  Running:

  Ran 2 tests with 0 failures and 0 errors in 1.144 seconds.
Running zope.testing.testrunner.layer.UnitTests tests:
  Tear down plone.restapi.testing.PloneRestApiDXPAMLayer:Integration in 0.000 seconds.
  Tear down plone.restapi.testing.PloneRestApiDXPAMLayer in 0.005 seconds.
  Tear down plone.app.contenttypes.testing.PloneAppContenttypes in 0.470 seconds.
  Tear down plone.app.event.testing.PAEventLayer in 0.007 seconds.
  Tear down plone.app.testing.layers.PloneFixture in 0.085 seconds.
  Tear down plone.testing.z2.Startup in 0.008 seconds.
  Tear down plone.testing.zca.LayerCleanup in 0.002 seconds.
  Set up zope.testing.testrunner.layer.UnitTests in 0.000 seconds.
  Running:

  Ran 58 tests with 0 failures and 0 errors in 0.029 seconds.
Tearing down left over layers:
  Tear down zope.testing.testrunner.layer.UnitTests in 0.000 seconds.
Total: 694 tests, 0 failures, 0 errors in 3 minutes 35.208 seconds.
```

**Comment by tisto** [comment]
Workaround here: https://github.com/plone/plone.restapi/pull/471

**Comment by lukasgraf** [comment]
@tisto I've been looking into this. Three observations so far:
1) It's definitely related to **freezing time**.
Asking around our office, we had a similar CI failure a while ago with a test that froze time (using `ftw.testing`'s freezer). Turned out, freezing the time wasn't really necessary in that test, so we were able to work around the issue without solving the underlying cause.  

2) It can be circumvented by setting the frozen time into the **future** instead of the past. 😕 

3) [ZODB seems to rely on `time.time()`](https://github.com/zopefoundation/ZODB/blob/master/src/ZODB/utils.py#L126-L131) to generate its **transaction IDs**. Pretty sure this is at the heart of the issue.

I'll look into this some more, though I first have to dig into the `ZODB` vs. `ZODB3` meta-egg confusion to figure out exactly what changed and what didn't.
